Add GetMastership function to mastership store

adibrastegarnia opened the pull request:
This PR adds a function to mastership store that we can get mastership term based on a given device id. 
Part of this issue https://github.com/onosproject/onos-config/issues/962

adibrastegarnia commented on pkg/store/mastership/store.go:
```diff
@@ -126,6 +131,15 @@ func (s *atomixStore) NodeID() cluster.NodeID {
+		return term, nil
+	}
+
+	return 0, fmt.Errorf("device mastership does not exist for device %s", deviceID)
```
@kuujo 
Can we assume the term starts from one and zero represents nothing or its initial value ?

kuujo replied:
The term for a real master is always `> 0`. The term will only be `0` if no master has ever been elected.

adibrastegarnia replied:
So it makes sense then.

kuujo commented on pkg/store/mastership/store.go:
```diff
@@ -37,6 +39,9 @@ type Store interface {
 	IsMaster(id topodevice.ID) (bool, error)
 
+	// GetTerm returns the mastership term for a given device
+	GetTerm(id topodevice.ID) (Term, error)
```
We might as well make this `GetMastership` instead of `GetTerm` so all the relevant information is returned in one step. Otherwise, mastership can change between `IsMaster` and `GetTerm`

adibrastegarnia replied:
Yeah. I was thinking about it... IsMaster function performs election as well which is a bit confusing. Suppose we have device1 and device2. If we call IsMaster for device1 then a master will be selected if it doesn't exist. So GetMastership performs election and returns all related information including Term assuming if a master doesn't exist for a device it picks one.

adibrastegarnia commented on pkg/store/mastership/election.go:
```diff
@@ -168,13 +169,13 @@ func (e *atomixDeviceMastershipElection) watchElection(ch <-chan *election.Event
+func (e *atomixDeviceMastershipElection) getMastership() *Mastership {
 	e.mu.RLock()
 	defer e.mu.RUnlock()
 	if e.mastership == nil || string(e.mastership.Master) != e.election.ID() {
```
I assume we still need to keep this condition

kuujo replied:
This condition should not be here. We're returning the mastership regardless of whether the local node is the master, but this check only returns the mastership if the local node is the master. Other nodes should have access to mastership info even if they're not the master.

adibrastegarnia replied:
Makes sense. So it should be moved to where it needs to be checked based on another comment.

kuujo replied:
Yeah... the local node ID can be read from the mastership store interface to check it.

adibrastegarnia commented on pkg/controller/filter.go:
```diff
@@ -43,11 +44,12 @@ func (f *MastershipFilter) Accept(id types.ID) bool {
-	if err != nil {
-		return false
+	master, err := f.Store.GetMastership(device)
+	if err == nil && master != nil {
```
@kuujo 
Please make sure it makes sense to you.

kuujo replied:
This should be checking whether the local node is the master. This check merely checks whether some master exists.

adibrastegarnia commented:
@kuujo 
I think I addressed your comments and see if you like the new changes. Please let me know if you have any more comments. I completely removed isMaster function and updated the places that is called instead of labeling it as Deprecated for now.

adibrastegarnia commented:
I should test it a bit more to make sure tests are passing without any problems.

adibrastegarnia commented on pkg/controller/snapshot/device/controller.go:
```diff
@@ -40,6 +42,7 @@ func NewController(mastership mastershipstore.Store, changes changestore.Store,
 		Resolver: &Resolver{
 			snapshots: snapshots,
 		},
+		NodeID: cluster.GetNodeID(),
```
NodeID injected that can be set by tests and here we use cluster.GetNodeID to get its value.

adibrastegarnia commented:
@kuujo 
I think it should be OK now.

kuujo commented on pkg/controller/filter.go:
```diff
@@ -35,6 +38,7 @@ type Filter interface {
 type MastershipFilter struct {
 	Store    mastershipstore.Store
 	Resolver DeviceResolver
+	NodeID   cluster.NodeID
```
We should probably make this private and add a `New` function to force it to be set correctly. Feels wrong to force other packages to set the node ID here.

adibrastegarnia replied:
@kuujo 
Agreed. Let me fix it.

adibrastegarnia replied:
I added a NewMastershipFilter function which accepts three arguments. The last argument is nodeID which can be empty and when it is empty it will be set to cluster NodeID otherwise it will be set to whatever it is passed to it (like what we have in filter_test.go). I also added a GetNodeID to filter that I can retrieve the private field in Accept function for checking local cluster node.

adibrastegarnia commented:
@kuujo 
I used POD_NAME for retrieving nodeID for now until we use cluster package in onos-lib-go for stores.  Perhaps I should do that one first before using mastership info for southbound. Looks like multiple store need to be updated to use that information accordingly.

kuujo commented on pkg/controller/filter.go:
```diff
@@ -35,6 +38,24 @@ type Filter interface {
+}
+
+// NewMastershipFilter creates an instance of mastershipFilter
+func NewMastershipFilter(store mastershipstore.Store, resolver DeviceResolver, nodeID cluster.NodeID) *MastershipFilter {
```
Actually, I think we should not have a `New*` function for the mastership filter. We should keep the API the same, but use `cluster.GetNodeID()` if the private `nodeID` field is empty.

adibrastegarnia replied:
@kuujo 
Done.
